From 533b989f98e8f9e6a03f582fb0a5ca2f81e209b3 Mon Sep 17 00:00:00 2001
From: Dimitri John Ledkov <dimitri.ledkov@surgut.co.uk>
Date: Wed, 31 Dec 2025 11:28:18 +0000
Subject: [PATCH] bazel: do not toy with FORTIFY_SOURCE settings

Multiple (re-)declarations of FORTIFY_SOURCE can cause errors and
build failures.

Our mongo builds are optimisized and declare FORTIFY_SOURCE in the
hardening-flags already, thus remove all attempts to set/unset
FORITY_SOURCE.
---
 .../mongo_apple_llvm_cc_toolchain_config.bzl  | 15 -----------
 bazel/toolchains/cc/mongo_defines.bzl         |  9 -------
 .../mongo_linux_cc_toolchain_config.bzl       | 15 -----------
 .../toolchains/mongo_cc_toolchain_config.bzl  | 15 -----------
 .../patches/17_revert_llvm_macos.path         | 26 +------------------
 .../toolchains/rbe_ubuntu2004/cc/BUILD        |  1 -
 .../rbe_ubuntu2004/cc/cc_toolchain_config.bzl | 15 -----------
 7 files changed, 1 insertion(+), 95 deletions(-)

diff --git a/bazel/toolchains/cc/mongo_apple/mongo_apple_llvm_cc_toolchain_config.bzl b/bazel/toolchains/cc/mongo_apple/mongo_apple_llvm_cc_toolchain_config.bzl
index fc1de0bb5ec..ae699c9d8af 100644
--- a/bazel/toolchains/cc/mongo_apple/mongo_apple_llvm_cc_toolchain_config.bzl
+++ b/bazel/toolchains/cc/mongo_apple/mongo_apple_llvm_cc_toolchain_config.bzl
@@ -318,21 +318,6 @@ def _impl(ctx):
         name = "default_compile_flags",
         enabled = True,
         flag_sets = [
-            flag_set(
-                actions = all_compile_actions,
-                flag_groups = [
-                    flag_group(
-                        # Security hardening requires optimization.
-                        # We need to undef it as some distributions now have it enabled by default.
-                        flags = ["-U_FORTIFY_SOURCE"],
-                    ),
-                ],
-                with_features = [
-                    with_feature_set(
-                        not_features = ["thin_lto"],
-                    ),
-                ],
-            ),
             flag_set(
                 actions = all_compile_actions,
                 flag_groups = ([
diff --git a/bazel/toolchains/cc/mongo_defines.bzl b/bazel/toolchains/cc/mongo_defines.bzl
index cb13c437476..36c8181b0df 100644
--- a/bazel/toolchains/cc/mongo_defines.bzl
+++ b/bazel/toolchains/cc/mongo_defines.bzl
@@ -51,14 +51,6 @@ ENTERPRISE_DEFINES = select({
     "//conditions:default": [],
 })
 
-# Fortify only possibly makes sense on POSIX systems, and we know that clang is
-# not a valid combination:
-# http://lists.llvm.org/pipermail/cfe-dev/2015-November/045852.html
-GCC_OPT_DEFINES = select({
-    "//bazel/config:gcc_opt": ["_FORTIFY_SOURCE=2"],
-    "//conditions:default": [],
-})
-
 DEBUG_DEFINES = select({
     "//bazel/config:dbg_enabled": [],
     "//conditions:default": ["NDEBUG"],
@@ -105,7 +97,6 @@ MONGO_GLOBAL_DEFINES = (
     UNDEFINED_SANITIZER_DEFINES +
     GLIBCXX_DEBUG_DEFINES +
     TCMALLOC_DEFINES +
-    GCC_OPT_DEFINES +
     BOOST_DEFINES +
     ABSEIL_DEFINES +
     PCRE2_DEFINES +
diff --git a/bazel/toolchains/cc/mongo_linux/mongo_linux_cc_toolchain_config.bzl b/bazel/toolchains/cc/mongo_linux/mongo_linux_cc_toolchain_config.bzl
index 710f8134608..2bacce072cb 100644
--- a/bazel/toolchains/cc/mongo_linux/mongo_linux_cc_toolchain_config.bzl
+++ b/bazel/toolchains/cc/mongo_linux/mongo_linux_cc_toolchain_config.bzl
@@ -135,21 +135,6 @@ def _impl(ctx):
         name = "default_compile_flags",
         enabled = True,
         flag_sets = [
-            flag_set(
-                actions = all_compile_actions,
-                flag_groups = [
-                    flag_group(
-                        # Security hardening requires optimization.
-                        # We need to undef it as some distributions now have it enabled by default.
-                        flags = ["-U_FORTIFY_SOURCE"],
-                    ),
-                ],
-                with_features = [
-                    with_feature_set(
-                        not_features = ["thin_lto"],
-                    ),
-                ],
-            ),
             flag_set(
                 actions = all_c_compile_actions,
                 flag_groups = [
diff --git a/bazel/toolchains/mongo_cc_toolchain_config.bzl b/bazel/toolchains/mongo_cc_toolchain_config.bzl
index a5c087ed00b..9395b007b3a 100644
--- a/bazel/toolchains/mongo_cc_toolchain_config.bzl
+++ b/bazel/toolchains/mongo_cc_toolchain_config.bzl
@@ -126,21 +126,6 @@ def _impl(ctx):
         name = "default_compile_flags",
         enabled = True,
         flag_sets = [
-            flag_set(
-                actions = all_compile_actions,
-                flag_groups = [
-                    flag_group(
-                        # Security hardening requires optimization.
-                        # We need to undef it as some distributions now have it enabled by default.
-                        flags = ["-U_FORTIFY_SOURCE"],
-                    ),
-                ],
-                with_features = [
-                    with_feature_set(
-                        not_features = ["thin_lto"],
-                    ),
-                ],
-            ),
             flag_set(
                 actions = all_c_compile_actions,
                 flag_groups = [
diff --git a/buildscripts/convert_scons_to_bazel/patches/17_revert_llvm_macos.path b/buildscripts/convert_scons_to_bazel/patches/17_revert_llvm_macos.path
index 1dce54de7a3..41d746bf121 100644
--- a/buildscripts/convert_scons_to_bazel/patches/17_revert_llvm_macos.path
+++ b/buildscripts/convert_scons_to_bazel/patches/17_revert_llvm_macos.path
@@ -323,7 +323,7 @@ new file mode 100644
 index 00000000000..8b31e5b60d9
 --- /dev/null
 +++ b/bazel/toolchains/cc/mongo_apple/mongo_apple_cc_toolchain_config.bzl
-@@ -0,0 +1,2798 @@
+@@ -0,0 +1,2774 @@
 +# Copyright 2019 The Bazel Authors. All rights reserved.
 +#
 +# Licensed under the Apache License, Version 2.0 (the "License");
@@ -2319,30 +2319,6 @@ index 00000000000..8b31e5b60d9
 +                flag_groups = [
 +                    flag_group(
 +                        flags = [
-+                            "-D_FORTIFY_SOURCE=1",
-+                        ],
-+                    ),
-+                ],
-+                with_features = [with_feature_set(not_features = ["asan"])],
-+            ),
-+            flag_set(
-+                actions = [
-+                    ACTION_NAMES.assemble,
-+                    ACTION_NAMES.preprocess_assemble,
-+                    ACTION_NAMES.linkstamp_compile,
-+                    ACTION_NAMES.c_compile,
-+                    ACTION_NAMES.cpp_compile,
-+                    ACTION_NAMES.cpp_header_parsing,
-+                    ACTION_NAMES.cpp_module_compile,
-+                    ACTION_NAMES.cpp_module_codegen,
-+                    ACTION_NAMES.lto_backend,
-+                    ACTION_NAMES.clif_match,
-+                    ACTION_NAMES.objc_compile,
-+                    ACTION_NAMES.objcpp_compile,
-+                ],
-+                flag_groups = [
-+                    flag_group(
-+                        flags = [
 +                            "-fstack-protector",
 +                            "-fcolor-diagnostics",
 +                            "-Wall",
diff --git a/src/third_party/grpc/dist/third_party/toolchains/rbe_ubuntu2004/cc/BUILD b/src/third_party/grpc/dist/third_party/toolchains/rbe_ubuntu2004/cc/BUILD
index a264ae3bd06..c4d0a99c554 100644
--- a/src/third_party/grpc/dist/third_party/toolchains/rbe_ubuntu2004/cc/BUILD
+++ b/src/third_party/grpc/dist/third_party/toolchains/rbe_ubuntu2004/cc/BUILD
@@ -119,7 +119,6 @@ cc_toolchain_config(
     opt_compile_flags = [
         "-g0",
         "-O2",
-        "-D_FORTIFY_SOURCE=1",
         "-DNDEBUG",
         "-ffunction-sections",
         "-fdata-sections",
diff --git a/src/third_party/grpc/dist/third_party/toolchains/rbe_ubuntu2004/cc/cc_toolchain_config.bzl b/src/third_party/grpc/dist/third_party/toolchains/rbe_ubuntu2004/cc/cc_toolchain_config.bzl
index 1a825f1fe17..c602837d740 100644
--- a/src/third_party/grpc/dist/third_party/toolchains/rbe_ubuntu2004/cc/cc_toolchain_config.bzl
+++ b/src/third_party/grpc/dist/third_party/toolchains/rbe_ubuntu2004/cc/cc_toolchain_config.bzl
@@ -202,21 +202,6 @@ def _impl(ctx):
         name = "default_compile_flags",
         enabled = True,
         flag_sets = [
-            flag_set(
-                actions = all_compile_actions,
-                flag_groups = [
-                    flag_group(
-                        # Security hardening requires optimization.
-                        # We need to undef it as some distributions now have it enabled by default.
-                        flags = ["-U_FORTIFY_SOURCE"],
-                    ),
-                ],
-                with_features = [
-                    with_feature_set(
-                        not_features = ["thin_lto"],
-                    ),
-                ],
-            ),
             flag_set(
                 actions = all_compile_actions,
                 flag_groups = ([
-- 
2.51.0

